src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -231,24 +311,56 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
             ArrayBuilder<(BoundReturnStatement, TypeSymbolWithAnnotations)> returnTypes,
             VariableState initialState,
             Action<BoundExpression, TypeSymbolWithAnnotations> callbackOpt)
+        {
+            return Analyze(compilation, method, node, diagnostics, useMethodSignatureReturnType, useMethodSignatureParameterTypes, methodSignatureOpt, returnTypes, initialState, analyzedNullabilityMapOpt: null, callbackOpt);
+        }
+
+        private static NullableWalker Analyze(
+            CSharpCompilation compilation,
+            MethodSymbol method,
+            BoundNode node,
+            DiagnosticBag diagnostics,
+            bool useMethodSignatureReturnType,
+            bool useMethodSignatureParameterTypes,
+            MethodSymbol methodSignatureOpt,
+            ArrayBuilder<(BoundReturnStatement, TypeSymbolWithAnnotations)> returnTypes,
+            VariableState initialState,
+            PooledDictionary<BoundExpression, TypeSymbolWithAnnotations> analyzedNullabilityMapOpt,
+            Action<BoundExpression, TypeSymbolWithAnnotations> callbackOpt)
         {
             Debug.Assert(diagnostics != null);
-            var walker = new NullableWalker(compilation, method, useMethodSignatureReturnType, useMethodSignatureParameterTypes, methodSignatureOpt, node, returnTypes, initialState, callbackOpt);
+            var walker = analyzedNullabilityMapOpt == null ?
+                new NullableWalker(compilation, method, useMethodSignatureReturnType, useMethodSignatureParameterTypes, methodSignatureOpt, node, returnTypes, initialState, callbackOpt) :
+                new NullableWalker(compilation, method, useMethodSignatureReturnType, useMethodSignatureParameterTypes, methodSignatureOpt, node, returnTypes, initialState, analyzedNullabilityMapOpt, callbackOpt);
             try
             {
                 bool badRegion = false;
                 ImmutableArray<PendingBranch> returns = walker.Analyze(ref badRegion);
                 diagnostics.AddRange(walker.Diagnostics);
+#if DEBUG
                 Debug.Assert(!badRegion);
+                // If the map was provided, we're in a nested call so we can't yet verify that all things are in the map.
+                if (analyzedNullabilityMapOpt == null)
+                {
+                    DebugVerifier.Verify(walker, node);
+                }
+#endif
+                return walker;
             }
             catch (BoundTreeVisitor.CancelledByStackGuardException ex) when (diagnostics != null)
             {
                 ex.AddAnError(diagnostics);
+                return null;
             }
-            finally
-            {
-                walker.Free();
-            }
+        }
+
+        private static BoundNode Rewrite(
+            NullableWalker walker,
+            BoundNode node)
+        {
+            var rewriter = new NullabilityRewriter(walker._analyzedNullabilityMap.ToImmutableDictionary());
```

Comment:
>walker._analyzedNullabilityMap.ToImmutableDictionary() [](start = 51, length = 54)

Consider having the caller of `Rewrite` pass in the `ImmutableDictionary<,>` rather than passing in a `NullableWalker` that needs freeing.